src/libAtomVM/mailbox.c

```diff
@@ -249,7 +253,7 @@ MailboxMessage *mailbox_process_outer_list(Mailbox *mbox)
             mbox->receive_pointer = previous_normal;
             // If we had a prev, set the prev's next to the new current.
             if (mbox->receive_pointer_prev) {
-                mbox->receive_pointer_prev->header.next = previous_normal;
+                mbox->receive_pointer_prev->next = previous_normal;
             } else if (mbox->inner_first == NULL) {
                 // If we had no first, this is the first message.
                 mbox->inner_first = previous_normal;
```

Comment:
I believe that the inner list (`inner_first` and `inner_last`) now only contains "normal" messages. So we may want to type these fields as `Message*` instead of `MailboxMessage*` to insist on this property.

Reply:
Sounds like a good point, so I rather replace it also in the mailbox itself.

Reply:
does this work for you ?
```c
@@ -109,11 +109,11 @@ typedef struct
     // Owner of mailbox perform deletion of all items (reversing the list into
     // the inner list)
     MailboxMessage *ATOMIC outer_first;
-    MailboxMessage *inner_first;
-    MailboxMessage *inner_last;
+    Message *inner_first;
+    Message *inner_last;
     // Receive pointers are on inner list items.
-    MailboxMessage *receive_pointer;
-    MailboxMessage *receive_pointer_prev;
+    Message *receive_pointer;
+    Message *receive_pointer_prev;
 } Mailbox;
```

Reply:
Sounds good, will help simplify some code.

Reply:
I'm not sure that I've got some simplified code here, the code here doesn't look compelling: https://github.com/atomvm/AtomVM/pull/483/commits/9dc99f7c2caf4b27997dd2e0ae0cc1c16fbcdb37

Reply:
Other problem:
```
In function ‘mailbox_destroy_signal_message’,
    inlined from ‘mailbox_destroy_message’ at /home/davide/Projects/AtomVM/b/feature-smp/src/libAtomVM/mailbox.c:80:5,
    inlined from ‘mailbox_remove’ at /home/davide/Projects/AtomVM/b/feature-smp/src/libAtomVM/mailbox.c:364:5:
/home/davide/Projects/AtomVM/b/feature-smp/src/libAtomVM/mailbox.c:73:13: warning: ‘free’ called on pointer ‘*mbox.receive_pointer’ with nonzero offset 24 [-Wfree-nonheap-object]
   73 |             free(m);
      |             ^~~~~~~
```

Reply:
> I'm not sure that I've got some simplified code here, the code here doesn't look compelling: [9dc99f7](https://github.com/atomvm/AtomVM/commit/9dc99f7c2caf4b27997dd2e0ae0cc1c16fbcdb37) Fun fact: I also did some kind of little mistake and I get a crash I cannot find, when I add some padding [1] to the structs in order to get some wrong assumption spotted.
> 
> [1]:
> 
> ```c
>  struct Message
>  {
> +    unsigned long pad[3];
> +
>      struct MailboxMessage base;
>  
>      int msg_memory_size;
> ```

Ok, found the first problem: that code does a lot of checks against null pointers and whatever, however when implementing the `Message *` inner list version, all the pointer must be checked  before getting any offset, eg.
```c
+    if (!ctx->mailbox.inner_first) {
+        return;
+    }
+
     MailboxMessage *msg = &ctx->mailbox.inner_first->base;
     while (msg) {
```

Ultimately I think that using `Message *` instead of `MailboxMessage *` makes the code way more complex.

Reply:
Sorry for the suggestion, let's not do it then.

Reply:
No worries, trying doing that it was worth it. I just reverted the change.